pandas/_testing/asserters.py

```diff
@@ -314,18 +314,16 @@ def _check_types(left, right, obj="Index") -> None:
             return
 
         assert_class_equal(left, right, exact=exact, obj=obj)
+        assert_attr_equal("inferred_type", left, right, obj=obj)
 
         # Skip exact dtype checking when `check_categorical` is False
-        if check_categorical:
-            assert_attr_equal("dtype", left, right, obj=obj)
-            if is_categorical_dtype(left.dtype) and is_categorical_dtype(right.dtype):
+        if is_categorical_dtype(left.dtype) and is_categorical_dtype(right.dtype):
```

Comment:
did this change any tests?

Reply:
I don't think so, right now. I think comparing dtype and inferred_type gives the same result for other dtypes that categoricals and numeric dtypes.

After #41153 we'll have int32 etc. dtypes, so then we'll have cases where it'll fail without this change.